File: test/helpers/ProcessHelpers.ts
/**
 * Helper functions for process and server management in tests.
 */

import { ChildProcess, spawn } from "child_process";
import { mkdir, rm } from "fs/promises";
import { tmpdir } from "os";
import { join } from "path";
import WebSocket from "ws";

// Constants
export const SERVER_PORT = 5580;
export const SERVER_WS_URL = `ws://localhost:${SERVER_PORT}/ws`;
export const DEVICE_PORT = 5540;
export const MANUAL_PAIRING_CODE = "34970112332";

/**
 * Creates temporary storage directories for server and device.
 * Returns paths that should be cleaned up after tests.
 */
export async function createTempStoragePaths(): Promise<{ serverStoragePath: string; deviceStoragePath: string }> {
    const timestamp = Date.now();
    const serverStoragePath = join(tmpdir(), `matter-test-server-${timestamp}`);
    const deviceStoragePath = join(tmpdir(), `matter-test-device-${timestamp}`);

    await mkdir(serverStoragePath, { recursive: true });
    await mkdir(deviceStoragePath, { recursive: true });

    return { serverStoragePath, deviceStoragePath };
}

/**
 * Cleans up temporary storage directories.
 */
export async function cleanupTempStorage(serverStoragePath: string, deviceStoragePath: string): Promise<void> {
    try {
        await rm(serverStoragePath, { recursive: true, force: true });
        await rm(deviceStoragePath, { recursive: true, force: true });
    } catch {
        // Ignore cleanup errors
    }
}

/**
 * Starts the Matter server process.
 */
export function startServer(storagePath: string): ChildProcess {
    const serverProcess = spawn(
        "node",
        ["--enable-source-maps", "packages/matter-server/dist/esm/MatterServer.js", `--storage-path=${storagePath}`],
        {
            cwd: process.cwd(),
            stdio: ["pipe", "pipe", "pipe"],
        },
    );

    serverProcess.stdout?.on("data", (data: Buffer) => {
        console.log("[server]", data.toString().trim());
    });
    serverProcess.stderr?.on("data", (data: Buffer) => {
        console.log("[server:err]", data.toString().trim());
    });

    return serverProcess;
}

/**
 * Starts the test device process.
 */
export function startTestDevice(storagePath: string): ChildProcess {
    return spawn("npx", ["tsx", "test/fixtures/TestLightDevice.ts", `--storage-path=${storagePath}`], {
        cwd: process.cwd(),
        stdio: ["pipe", "pipe", "pipe"],
    });
}

/**
 * Waits for a port to be ready by attempting WebSocket connections.
 */
export async function waitForPort(port: number, timeoutMs = 30_000): Promise<void> {
    const start = Date.now();
    while (Date.now() - start < timeoutMs) {
        try {
            const ws = new WebSocket(`ws://localhost:${port}/ws`);
            await new Promise<void>((resolve, reject) => {
                ws.on("open", () => {
                    ws.close();
                    resolve();
                });
                ws.on("error", reject);
            });
            return;
        } catch {
            await new Promise(r => setTimeout(r, 500));
        }
    }
    throw new Error(`Timeout waiting for port ${port}`);
}

/**
 * Waits for the device process to output readiness indicators.
 */
export function waitForDeviceReady(process: ChildProcess, timeoutMs = 30_000): Promise<void> {
    return new Promise((resolve, reject) => {
        const timeout = setTimeout(() => {
            reject(new Error("Timeout waiting for device to be ready"));
        }, timeoutMs);

        const onData = (data: Buffer) => {
            const output = data.toString();
            console.log("[device]", output.trim());
            // Device is ready when it shows the pairing codes
            if (output.includes("Manual pairing code:") || output.includes("commissioned")) {
                clearTimeout(timeout);
                process.stdout?.off("data", onData);
                // Give it a moment to fully initialize network
                setTimeout(resolve, 2000);
            }
        };

        process.stdout?.on("data", onData);
        process.stderr?.on("data", (data: Buffer) => {
            console.log("[device:err]", data.toString().trim());
        });
    });
}

/**
 * Gracefully kills a process.
 */
export async function killProcess(process: ChildProcess | undefined): Promise<void> {
    if (process) {
        process.kill("SIGTERM");
        await new Promise(r => setTimeout(r, 1000));
    }
}

File: python_client/tests/helpers/process_helpers.py
"""Process management helpers for integration tests.

Ports the functionality of the JS ProcessHelpers.ts for use in Python integration tests.
Provides helpers to start/stop the Matter.js server and test device subprocesses.
"""

from __future__ import annotations

import asyncio
import os
import shutil
import signal
import subprocess
import tempfile
import time
from pathlib import Path

import aiohttp

REPO_ROOT = Path(__file__).parent.parent.parent.parent  # python_client/tests/helpers -> repo root
SERVER_PORT = 5580
SERVER_WS_URL = f"ws://localhost:{SERVER_PORT}/ws"
DEVICE_PORT = 5540
MANUAL_PAIRING_CODE = "34970112332"


def create_temp_storage_paths() -> tuple[str, str]:
    """Create temporary storage directories for server and device.

    Returns a tuple of (server_storage_path, device_storage_path).
    Both directories are created on disk immediately.
    """
    timestamp = int(time.time() * 1000)
    server_path = os.path.join(tempfile.gettempdir(), f"matter-test-server-{timestamp}")
    device_path = os.path.join(tempfile.gettempdir(), f"matter-test-device-{timestamp}")
    os.makedirs(server_path, exist_ok=True)
    os.makedirs(device_path, exist_ok=True)
    return server_path, device_path


def cleanup_temp_storage(server_path: str, device_path: str) -> None:
    """Remove temporary storage directories.

    Silently ignores errors if directories don't exist or can't be removed.
    """
    for path in (server_path, device_path):
        try:
            shutil.rmtree(path)
        except OSError:
            pass


def start_server(storage_path: str) -> subprocess.Popen:
    """Start the Matter.js server process.

    Args:
        storage_path: Path to the temporary storage directory for server state.

    Returns:
        The subprocess.Popen handle for the running server.
    """
    server_script = REPO_ROOT / "packages" / "matter-server" / "dist" / "esm" / "MatterServer.js"
    return subprocess.Popen(
        [
            "node",
            "--enable-source-maps",
            str(server_script),
            f"--storage-path={storage_path}",
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        cwd=str(REPO_ROOT),
    )


def start_test_device(storage_path: str) -> subprocess.Popen:
    """Start the test light device process.

    Args:
        storage_path: Path to the temporary storage directory for device state.

    Returns:
        The subprocess.Popen handle for the running test device.
    """
    device_script = (
        REPO_ROOT / "packages" / "matter-server" / "test" / "fixtures" / "TestLightDevice.ts"
    )
    return subprocess.Popen(
        [
            "npx",
            "tsx",
            str(device_script),
            f"--storage-path={storage_path}",
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        cwd=str(REPO_ROOT),
    )


async def wait_for_port(port: int, timeout: float = 30.0) -> None:
    """Wait for the WebSocket server to accept connections on the given port.

    Repeatedly attempts a WebSocket connection until successful or timeout.

    Args:
        port: The port number to connect to.
        timeout: Maximum time to wait in seconds.

    Raises:
        TimeoutError: If the server doesn't become available within the timeout.
    """
    start = time.monotonic()
    while time.monotonic() - start < timeout:
        try:
            async with aiohttp.ClientSession() as session:
                async with session.ws_connect(f"ws://localhost:{port}/ws", timeout=2):
                    return
        except (aiohttp.ClientError, OSError, asyncio.TimeoutError):
            await asyncio.sleep(0.5)
    raise TimeoutError(f"Timeout waiting for WebSocket server on port {port}")


async def wait_for_device_ready(process: subprocess.Popen, timeout: float = 30.0) -> None:
    """Wait for the test device to be ready by monitoring its stdout.

    Watches for the "Manual pairing code:" or "commissioned" output lines
    that indicate the device is ready to accept connections.

    Args:
        process: The device subprocess to monitor.
        timeout: Maximum time to wait in seconds.

    Raises:
        RuntimeError: If the device process exits unexpectedly.
        TimeoutError: If the device doesn't become ready within the timeout.
    """
    start = time.monotonic()
    assert process.stdout is not None

    while time.monotonic() - start < timeout:
        # Non-blocking readline
        line = process.stdout.readline()
        if not line:
            if process.poll() is not None:
                raise RuntimeError("Device process exited unexpectedly")
            await asyncio.sleep(0.1)
            continue
        line = line.strip()
        if "Manual pairing code:" in line or "commissioned" in line.lower():
            # Give the device's network stack time to fully initialize
            await asyncio.sleep(2)
            return

    raise TimeoutError("Timeout waiting for device to be ready")


def kill_process(process: subprocess.Popen | None, timeout: float = 5.0) -> None:
    """Gracefully kill a process, falling back to SIGKILL if needed.

    Sends SIGINT first to allow graceful shutdown, then SIGKILL if the
    process doesn't exit within the timeout.

    Args:
        process: The process to kill, or None (no-op).
        timeout: Time to wait for graceful shutdown before forcing kill.
    """
    if process is None or process.poll() is not None:
        return
    process.send_signal(signal.SIGINT)
    try:
        process.wait(timeout=timeout)
    except subprocess.TimeoutExpired:
        process.kill()
        process.wait(timeout=2)
